// @progress/kendo-angular-charts/dist/npm/common/theme.service.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var core_1 = require('@angular/core');
var configuration_service_1 = require('./configuration.service');
/**
 * @hidden
 */
var ThemeService = (function (_super) {
    __extends(ThemeService, _super);
    function ThemeService() {
        _super.apply(this, arguments);
        this.loaded = false;
        this.loading = false;
    }
    ThemeService.prototype.isLoaded = function () {
        return this.loaded;
    };
    ThemeService.prototype.markLoaded = function () {
        this.loaded = true;
    };
    ThemeService.prototype.startedLoading = function () {
        this.loading = true;
    };
    ThemeService.prototype.isLoading = function () {
        return this.loading;
    };
    ThemeService.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    ThemeService.ctorParameters = function () { return []; };
    return ThemeService;
}(configuration_service_1.ConfigurationService));
exports.ThemeService = ThemeService;
