// @progress/kendo-angular-charts/dist/es/common/theme.service.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
import { Injectable } from '@angular/core';
import { ConfigurationService } from './configuration.service';
/**
 * @hidden
 */
export var ThemeService = (function (_super) {
    __extends(ThemeService, _super);
    function ThemeService() {
        _super.apply(this, arguments);
        this.loaded = false;
        this.loading = false;
    }
    ThemeService.prototype.isLoaded = function () {
        return this.loaded;
    };
    ThemeService.prototype.markLoaded = function () {
        this.loaded = true;
    };
    ThemeService.prototype.startedLoading = function () {
        this.loading = true;
    };
    ThemeService.prototype.isLoading = function () {
        return this.loading;
    };
    ThemeService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    ThemeService.ctorParameters = function () { return []; };
    return ThemeService;
}(ConfigurationService));
